crates/matrix-sdk/src/event_cache/pagination.rs

```diff
@@ -121,126 +121,162 @@ impl RoomPagination {
     async fn run_backwards_impl(&self, batch_size: u16) -> Result<Option<BackPaginationOutcome>> {
         const DEFAULT_WAIT_FOR_TOKEN_DURATION: Duration = Duration::from_secs(3);
 
-        let prev_token = self.get_or_wait_for_token(Some(DEFAULT_WAIT_FOR_TOKEN_DURATION)).await;
-
-        let prev_token = match prev_token {
-            PaginationToken::HasMore(token) => Some(token),
-            PaginationToken::None => None,
-            PaginationToken::HitEnd => {
-                debug!("Not back-paginating since we've reached the start of the timeline.");
-                return Ok(Some(BackPaginationOutcome { reached_start: true, events: Vec::new() }));
+        // First off, remember that's the `RoomEvents` might be partially loaded
+        // (because not all events are fully loaded).
+        //
+        // Knowing that, let's try to load more events from the storage. If there is
+        // nothing from the storage, let's fallback on the network.
+        {
+            let mut state = self.inner.state.write().await;
+
+            // Try to load one chunk backwards. If it returns `Some(_)`, it means new events
+            // have been inserted. No need to reach the network!
+            if let Some(((events, reached_start), sync_timeline_events_diffs)) =
+                state.load_more_events_backwards().await?
+            {
+                if !sync_timeline_events_diffs.is_empty() {
+                    let _ = self.inner.sender.send(RoomEventCacheUpdate::UpdateTimelineEvents {
+                        diffs: sync_timeline_events_diffs,
+                        origin: EventsOrigin::Pagination,
+                    });
+                }
+
+                return Ok(Some(BackPaginationOutcome {
+                    reached_start,
+                    // This is a backwards pagination. `BackPaginationOutcome` expects events to be
+                    // in “reverse order”.
+                    events: events.into_iter().rev().collect(),
+                }));
             }
-        };
+        }
 
-        let paginator = &self.inner.paginator;
+        // Now we can fallback onto the network.
+        {
```

Comment:
Can we avoid opening a new scope here? I don't find it useful, since it's closed at the same time as the function's body. (Only makes the diff harder to read :woozy_face:)